core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -970,6 +970,11 @@ protected void runTask(Task task) throws Exception {
             @Override
             protected void onInitMilestoneAttained(InitMilestone milestone) {
                 initLevel = milestone;
+                if (milestone==PLUGINS_PREPARED) {
+                    // set up Guice to enable injection as early as possible
+                    // before this milestone, ExtensionList.ensureLoaded() won't actually try to locate instances
```

Comment:
I'm not sure if I understand. This hunk does have a hackish feel to it, but that's because of our two minds about Guice; on one hand it's "just" an `ExtensionFinder` like every other, yet it's also a critical dependency as can be seen in `Jenkins.getInjector()`

I don't think JENKINS-22050 is a problem, but even if it is, I'd think this change is going to make that fix easier by allowing the reactor to reason about what initializer methods need.